File: cpp/graphAlgorithms/misc/unit/testBfs.cpp
/** @file testBfs.cpp
 * 
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "stdinc.h"
#include "Adt.h"
#include "Util.h"
#include "List.h"
#include "Graph.h"

namespace grafalgo {
extern void bfs(Graph&, vertex, List&);
}

using namespace grafalgo;

/** usage: testBfs
 *  
 *  TestBfs reads a graph from stdin, and lists its vertices in
 *  breadth-first order starting from vertex 1.
 */
int main() {
	Graph g; cin >> g;
	List vlist(g.n());
	bfs(g,1,vlist);
	cout << g << vlist << endl;
}

File: js/graphAlgorithms/misc/unit/testBfs.mjs
/** \file testBfs.java
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import { assert, AssertFail } from '../../../common/Assert.mjs';
import { matches, Mismatch } from '../../../common/Testing.mjs';
import bfs from '../bfs.mjs';
import Graph from '../../../dataStructures/graphs/Graph.mjs';

try {
	console.log("testing bfs");

	let n = 6; let g = new Graph(n);
	g.fromString("{a[b d e] b[a c f] c[b d f] d[a c e] e[a d] f[b c]}");

	let vlist = bfs(g, 1);
	matches(g.ilist2string(vlist), '[a b d e c f]', 'a1');
	vlist = bfs(g, 4);
	matches(g.ilist2string(vlist), '[d a c e b f]', 'a2');
	vlist = bfs(g, 6);
	matches(g.ilist2string(vlist), '[f b c a d e]', 'a3');

} catch (e) {
    if (e instanceof Mismatch) {
        console.log(e.name + ': ' + e.message);
	} else if (e instanceof AssertFail) {
		console.error(e.stack);
	} else {
        throw(e);
	}
}
